socorro/unittest/cron/jobs/test_update_signatures.py

```diff
@@ -0,0 +1,257 @@
+
+
+class FakeModel(object):
+    def __init__(self):
+        self._get_steps = []
+
+    def add_get_step(self, assertions, response):
+        assertions = assertions or []
+        self._get_steps.append({
+            'assertions': assertions,
+            'response': response
+        })
+
+    def get(self, *args, **kwargs):
+        if not self._get_steps:
+            raise Exception('Unexpected call to .get()')
+
+        step = self._get_steps.pop(0)
+
+        for assertion in step['assertions']:
+            assertion(args, kwargs)
+        return step['response']
+
+
+class UpdateSignaturesCronAppTestCase(IntegrationTestBase):
+    def _setup_config_manager(self):
+        return super(UpdateSignaturesCronAppTestCase, self)._setup_config_manager(
+            jobs_string='socorro.cron.jobs.update_signatures.UpdateSignaturesCronApp|1h'
+        )
+
+    def _truncate_signatures(self):
+        """Wipe the signatures table"""
+        self.conn.cursor().execute('TRUNCATE signatures CASCADE')
+        self.conn.commit()
+
+    def fetch_signatures_data(self):
+        cursor = self.conn.cursor()
+        cursor.execute("SELECT signature, first_build, first_report FROM signatures")
+        results = cursor.fetchall()
+        return [
+            {
+                'signature': str(result[0]),
+                'first_build': str(result[1]),
+                'first_report': str(result[2]),
+            }
+            for result in results
+        ]
+
+    def setUp(self):
+        super(UpdateSignaturesCronAppTestCase, self).setUp()
+        self._truncate_signatures()
+
+    def tearDown(self):
+        super(UpdateSignaturesCronAppTestCase, self).tearDown()
+        self._truncate_signatures()
+
+    @mock.patch('socorro.cron.jobs.update_signatures.SuperSearch')
+    def test_no_crashes_to_process(self, mock_supersearch):
+        supersearch = FakeModel()
+        mock_supersearch.return_value = supersearch
+
+        # Mock SuperSearch to return no results
+        supersearch.add_get_step(
+            assertions=None,
+            response={
+                'errors': [],
+                'hits': [],
+                'total': 0,
+                'facets': {}
+            }
+        )
+
+        # Run crontabber
+        config_manager = self._setup_config_manager()
+        with config_manager.context() as config:
+            crontabberapp = CronTabberApp(config)
+            crontabberapp.run_one('update-signatures')
+
+        # Assert the job ran correctly
+        crontabber_info = self._load_structure()
+        assert crontabber_info['update-signatures']['last_error'] == {}
+        assert crontabber_info['update-signatures']['last_success']
+
+        # Assert that nothing got inserted
+        data = self.fetch_signatures_data()
+        assert len(data) == 0
+
+    @mock.patch('socorro.cron.jobs.update_signatures.SuperSearch')
+    def test_signature_insert_and_update(self, mock_supersearch):
```

Comment:
This doesn't look parameterizable since it needs to insert and then update, but the "run job" and "assert no failures" bits could be factored out into methods to make the test a bit more concise.

Reply:
I see where you're coming from, but it's two tests and they're not that long. The bulk of the code in both tests is Python data structures that would be there in either case.

I think I'd rather leave them as they are now unless you feel very strongly about this. If we add more tests in the future, we can spend time reducing repeated bits because we'll have more repetition then.

Reply:
Ahh... ok. I forgot about the `WHERE NOT EXISTS` part. I couldn't figure out why it seemed like only one executed based on the final return codes. That makes soooo much sense now.